src/controller/Controller.js

```diff
@@ -0,0 +1,257 @@
+import { GiftService, OrderService, DiscountService, BadgeService } from '../service/index.js';
+
+import InputView from '../InputView.js';
+import OutputView from '../OutputView.js';
+
+import { Badge, Receipt } from '../domain/index.js';
+
+import SYSTEM from '../constants/system.js';
+
+/**
+ * @typedef {import('../service/DiscountService.js').BenefitResult} BenefitResult
+ */
+
+class Controller {
+  /**
+   * 입출력을 담당하는 View입니다.
+   */
+  #view = {
+    input: InputView,
+    output: OutputView,
+  };
+
+  /**
+   * 비즈니스 로직을 담당하는 Service입니다.
+   */
+  #service = {
+    order: OrderService,
+    gift: GiftService,
+    discount: DiscountService,
+    badge: BadgeService,
+  };
+
+  /**
+   * 할인 계산 프로그램을 시작합니다.
+   */
+  async start() {
+    this.#printStartComment();
+
+    const receipt = await this.#createReceipt();
+    await this.#processOrder(receipt);
+
+    const benefits = this.#processAllEvents(receipt);
+    this.#printBenefits(benefits);
+    this.#printBenefitAmount(receipt);
+    this.#printPaymentAmount(receipt);
+
+    const badge = this.#processBadgeEvent(receipt);
+    this.#printBadge(badge);
+  }
+
+  /**
+   * 방문 일자를 기록한 영수증을 발행합니다.
+   * @returns {Promise<Receipt>} - 발행된 영수증입니다.
+   */
+  async #createReceipt() {
+    let receipt;
+
+    await this.#handleError(async () => {
+      const date = await this.#readVisitDate();
+      receipt = this.#service.order.publishReceipt(date);
+    });
```

Comment:
```suggestion
    const receipt = await this.#handleError(async () => {
      const date = await this.#readVisitDate();
      return this.#service.order.publishReceipt(date);
    });
```
```js
 async #handleError(action) {
    try {
      return await action();
    } catch ({ message }) {
      this.#view.output.error(message);
      await this.#handleError(action);
    }
  }
```

p5 - 이런 식으로 사용하면 let 사용을 줄일 수 있고 순수 함수 형태로 전개 되기 때문에 사이드 이펙트를 줄일 수 있을거 같아 제안 드려봅니다!